controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -259,6 +268,41 @@ func (r *BareMetalHostReconciler) Reconcile(request ctrl.Request) (result ctrl.R
 	return
 }
 
+// Consume inspect.metal3.io/hardwaredetails when either
+// inspect.metal3.io=disabled or there are no existing HardwareDetails
+func (r *BareMetalHostReconciler) updateHardwareDetails(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (bool, error) {
+	reqLogger := r.Log.WithValues("baremetalhost", request.NamespacedName)
+	updated := false
+	if host.Status.HardwareDetails == nil || inspectionDisabled(host) {
+		objHardwareDetails, err := r.getHardwareDetailsFromAnnotation(host)
+		if err != nil {
+			reqLogger.Info("Error getting HardwareDetails from annotation")
+			return updated, err
+		} else if objHardwareDetails != nil {
+			reqLogger.Info("Setting HardwareDetails from annotation")
+			host.Status.HardwareDetails = objHardwareDetails
+			err = r.saveHostStatus(host)
+			if err != nil {
+				reqLogger.Info("Could not update hardwaredetails from annotation")
+				return updated, err
+			}
+			updated = true
+		}
+	}
+	// We either just processed the annotation, or the status is already set
+	// so we remove it
```

Comment:
Based on https://github.com/metal3-io/baremetal-operator/pull/796/files#r585936511, do we need to remove the annotation at all?

Reply:
I copied this behavior from the status annotation code https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L145..L153

However in that case the rationale for removing is that the status data will become outdated, which isn't the case here.

However, we do need some way to handle the following scenarios:

1. hardwaredetails annotation is set, but we ignore it (inspection is enabled, and status.hardware already set)
2. User updates hardwaredetails annotation in the case where inspection is disabled

For (1) we should perhaps remove the annotation and generate an event, rather than just ignoring it forever?

And for (2) we have to either remove the annotation (so we can detect new data because the annotation got re-applied), or we need to compare old/new status (which isn't impossible, but it'll mean json encoding the existing status and comparing the strings, which seems more complex and error prone than just removing the annotation after we consume it?)

Reply:
I'll resolve this thread, since we covered generating an events in the previous one and I added an update there